gateway/test/src/test/java/io/apiman/gateway/test/junit/vertx3/ApiToFileRegistry.java

```diff
@@ -0,0 +1,186 @@
+import io.vertx.core.file.FileSystem;
+import io.vertx.core.json.Json;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+@SuppressWarnings("nls")
+public class ApiToFileRegistry extends InMemoryRegistry {
+
+    private FileSystem fileSystem;
+    private File file;
+    private JsonArray clients = new JsonArray();
+    private JsonArray apis = new JsonArray();
+    private JsonObject root = new JsonObject(); private void linkRoot() {
+        root.put("clients", clients);
+        root.put("apis", apis);
+    }
+    private Map<Client, JsonObject> clientMap = new LinkedHashMap<>();
+    private Map<Api, JsonObject> apiMap = new LinkedHashMap<>();
+    private EventBus eb;
+
+    public ApiToFileRegistry(Vertx vertx, IEngineConfig foo, Map<String, String> config) {
+        super();
+        this.eb = vertx.eventBus();
+        this.fileSystem = vertx.fileSystem();
+        linkRoot();
+        createTempFile();
+        createResetListener();
+    }
+
+    private void createResetListener() {
+        eb.consumer(ApiToFileRegistry.class.getCanonicalName(), msg -> {
+           if (msg.headers().get("action").equals("reset")) {
+               System.out.println("Resetting!");
+               reset();
+           } else {
+               throw new RuntimeException("Unknown action in " + msg.headers());
+           }
+        });
+    }
+
+    private void createTempFile()  {
+        try {
+            file = File.createTempFile("apiman", "fs");
+            System.setProperty("CONFIG_FILE_PATH", "file:///" + file.getAbsolutePath());
+            System.out.println("file path file:///" + file.getAbsolutePath());
+            file.deleteOnExit();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public void reset() {
+        fileSystem.deleteBlocking(file.getAbsolutePath());
+        super.getMap().clear();
+        apis.clear();
+        clients.clear();
+        clientMap.clear();
+        apiMap.clear();
+        root.clear();
+        linkRoot();
+    }
+
+    private void rewrite() {
+        fileSystem.writeFileBlocking(file.getAbsolutePath(), Buffer.buffer(root.encodePrettily()));
+        restartReader();
+    }
+
+    private void restartReader() {
+        eb.publish("restartReader", "reset!");
+    }
+
+    private void publish(Api api) {
+        System.out.println("WANT TO PUBLISH " + api);
+        JsonObject apiJson = new JsonObject(Json.encode(api));
+        apis.add(apiJson);
+        apiMap.put(api, apiJson);
+        rewrite();
+    }
+
+    private void retire(Api api) {
+        JsonObject removeJson = apiMap.remove(api);
+        apis.remove(removeJson);
+        rewrite();
+    }
+
+    private void register(Client client) {
+        JsonObject clientJson = new JsonObject(Json.encode(client));
```

Comment:
These methods use the InMemoryRegistry to do validation, and then rewrite the JSON file when necessary.